### PR TITLE
Incorporate Singletons to Hyrise object

### LINKED ISSUE #1786: Add missing singletons to the Hyrise object
## CurrentScheduler

I was [wrong](https://github.com/hyrise/hyrise/pull/1761/files#r311565893) about the CurrentScheduler. While `static std::shared_ptr<AbstractScheduler> _instance` does not make it a singleton, it almost behaves like one. We should add it to `Hyrise` so that it gets reset when *everything* is supposed to be reset.

It might make more sense to add a `NoScheduler`, `ImmediateExecutionScheduler` (or similar) which takes the place of the scheduler if none is set.

## Others
* Topology
* JitRepository
* Console
* BenchmarkRunner (#1721)

### PR DESCRIPTION
**Pull request by d-justen**
fixes #1786

### CODE REVIEW

File: src/lib/tasks/server/execute_server_prepared_statement_task.cpp [hunk 1]
```diff
@@ -10,7 +11,7 @@ namespace opossum {
 void ExecuteServerPreparedStatementTask::_on_execute() {
   try {
     const auto tasks = OperatorTask::make_tasks_from_operator(_prepared_plan, CleanupTemporaries::Yes);
-    CurrentScheduler::schedule_and_wait_for_tasks(tasks);
+    Hyrise::get().current_scheduler.schedule_and_wait_for_tasks(tasks);
```
Comment (mrks) [inline src/lib/tasks/server/execute_server_prepared_statement_task.cpp]: Now that we access it through `Hyrise`, does "current" bring anything to the table?

File: src/lib/scheduler/current_scheduler.hpp [hunk 2]
```diff
@@ -14,34 +15,38 @@ class AbstractScheduler;
 /**
  * Holds the singleton instance (or the lack of one) of the currently active Scheduler
  */
-class CurrentScheduler {
+class CurrentScheduler : public Noncopyable {
```
Comment (mrks) [inline src/lib/scheduler/current_scheduler.hpp]: What is the point of this class? Now that we have `NoScheduler`, all those `if (_instance)` checks should be gone, anyway. Can we get rid of this indirection?

File: src/lib/scheduler/no_scheduler.hpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include "abstract_scheduler.hpp"
+#include "abstract_task.hpp"
+
+namespace opossum {
+
+class NoScheduler : public AbstractScheduler {
```
Comment (mrks) [inline src/lib/scheduler/no_scheduler.hpp]: Naming: `ImmediateExecutionScheduler`? I'll admit that it is bulky, maybe you have a better idea?

Comment (mrks) [inline src/lib/scheduler/no_scheduler.hpp]: doc

File: src/lib/scheduler/no_scheduler.hpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+  void finish() override {}
+
+  bool active() const override { return false; }
+
+  const std::vector<std::shared_ptr<TaskQueue>>& queues() const override { return _queues; }
+
+  void schedule(std::shared_ptr<AbstractTask> task, NodeID preferred_node_id = CURRENT_NODE_ID,
+                SchedulePriority priority = SchedulePriority::Default) override {
+    if (task->is_ready()) task->execute();
```
Comment (mrks) [inline src/lib/scheduler/no_scheduler.hpp]: code in cpp

File: src/lib/hyrise.hpp [hunk 5]
```diff
@@ -2,26 +2,42 @@
...
-  static void reset() { get() = Hyrise{}; }
+  // You should be very sure that this is what you want.
+  static void reset() {
+    Hyrise::get().current_scheduler.get()->finish();
+    get() = Hyrise{};
+  }
+
+  void set_jit_repository(std::shared_ptr<JitRepository> repo) { jit_repository = repo; }
+
+  void set_benchmark_runner(std::shared_ptr<BenchmarkRunner> runner) { benchmark_runner = runner; }
```
Comment (mrks) [inline src/lib/hyrise.hpp]: Why do we need setters if the members are writable anyway?

File: src/lib/hyrise.hpp [hunk 6]
```diff
@@ -2,26 +2,42 @@
...
 
   PluginManager plugin_manager;
   StorageManager storage_manager;
   TransactionManager transaction_manager;
   MetaTableManager meta_table_manager;
+  Topology topology;
+  CurrentScheduler current_scheduler;
+
+  std::shared_ptr<JitRepository> jit_repository;
+  std::shared_ptr<BenchmarkRunner> benchmark_runner;
```
Comment (mrks) [inline src/lib/hyrise.hpp]: The Hyrise object should not be a co-owner of the benchmark runner. If whoever created it decides that it is done with it, it should be deleted. weak_ptr?

**Comment by d-justen** [comment]
Does anybody know what this clangDebugMac error is about? It says:
```console
In file included from /Users/hyrise-ci/.jenkins/workspace/hyrise_hyrise_PR-1812/src/lib/scheduler/current_scheduler.hpp:8:

/Users/hyrise-ci/.jenkins/workspace/hyrise_hyrise_PR-1812/src/lib/utils/tracing/probes.hpp:23:10: fatal error: 'provider.hpp' file not found

#include "provider.hpp"

         ^~~~~~~~~~~~~~

1 error generated.

```

I don't really see how this is connected to this PR specifically.

**Pull request merged**